### DISCUSSION TITLE
Revisando o que aprendemos

### DISCUSSION THREAD

**Discussion by ThiagoBussola** [TI-UNICESUMAR/2024-topicos-em-programacao-ESOFT7S#3] [Node.js]
Com suas palavras, crie um resumo e explique os seguintes tópicos:

- O que é um runtime
- Componentes do runtime JS Node.js
- O que é e o que faz a V8
- O que é e o que faz a Libuv
- O que é assincronia
- O que são callbacks
- Como funciona o Loop de eventos do Node.js e quais são seus componentes
- Qual a diferença do Loop de Eventos do Node.js para o Loop de eventos utilizados nos Browsers

**Comment by richwrd** [comment]
[eduardorichard.zip](https://github.com/TI-UNICESUMAR/2024-topicos-em-programacao-ESOFT7S/files/14303454/eduardorichard.zip)

**Comment by lucascossitt** [comment]
Lucas Zuin Cossitt - Ra: 21009453-2

Nos PDFs, exploramos como as APIs facilitam conversas entre softwares e mergulhamos no universo do Node.js para entender como ele torna a web mais ágil. A primeira parte é um guia sobre a comunicação web, explicando como os dados viajam e são trocados. A segunda parte é uma viagem ao coração do Node.js, revelando como ele lida com múltiplas solicitações simultaneamente, mantendo tudo rápido e responsivo. É como aprender a linguagem secreta da internet e, ao mesmo tempo, descobrir o motor que mantém tudo rodando suavemente.

**Comment by NatanColleoni** [comment]
- O que é um runtime?
    - É o tempo de execução de um programa, fornece um conjunto de recursos e bibliotecas, como ambientes virtuais, gerenciamento de memória, entre outras funcionalidades.
- Componentes do runtime JS Node.js
    - Motor V8
    - Libuv
    - Camada C++
- O que é e o que faz a V8
    - V8 é a engine do google que permite que código javascript seja interpretado e executado.
    - Pega uma string com código javascript, valida, e o executa em C++.
- O que é e o que faz a Libuv
    - É uma biblioteca C, que nos permite trabalhar com entrada/saida de dados sem bloquear a thread principal.
- O que é assincronia
    - É uma técnica que permite que o programa execute tarefas que podem demorar (acesso ao banco de dados, chamadas externas), sem bloquear o fluxo do programa e permitindo que outras operações sejam realizadas.
- O que são callbacks
    - É uma função passada como pârametro para outra função, para que seja executada após alguma ação.  Por exemplo, setTimeout recebe uma função, e após passar o tempo definido, esta função é executada.
- Como funciona o Loop de eventos do Node.js e quais são seus componentes
    - O loop de eventos monitora e gerencia a ordem de execução de funções, ela verifica se há itens assincronos para serem executados, colocam eles para rodarem na libuv que então os retorna para a fila de eventos, o event loop irá pegar o resultado da libuv e irá inserir novamente na call stack. É composto pela call stack, libuv e event queue e thread pool.
- Qual a diferença do Loop de Eventos do Node.js para o Loop de eventos utilizados nos Browsers
    - A principal diferença é que o loop de eventos do browser é utilizado para tratar elementos do DOM, como interações do usuário. Enquanto o loop de eventos do node.js é utilizado para operações de entrada e saída, como escrita no banco de dados e chamadas externas.
    - No loop de eventos do browser, ao invés das operações serem executadas em um servidor node, elas são executadas através de uma Web API.

**Comment by Gus-hcs** [comment]
Runtime: Imagine um palco onde o JavaScript pode atuar fora do navegador; esse é o papel do runtime no Node.js, fornecendo todo o cenário necessário.

Motor V8: Como um tradutor ágil, o V8 converte JavaScript em linguagem de máquina, acelerando a execução dos scripts.

Libuv: Essa biblioteca é a chave para fazer várias coisas ao mesmo tempo, permitindo que o Node.js lide com várias tarefas sem travar.

Assincronia: É como ler um livro enquanto cozinha; você não precisa parar uma atividade para fazer a outra, maximizando a eficiência do seu tempo.

Callbacks: Imagine delegar tarefas e ser notificado quando elas estiverem prontas; os callbacks são essas notificações no mundo do código.

Loop de Eventos: Funciona como um gerente que organiza e executa tarefas baseado em sua prioridade e disponibilidade, mantendo o aplicativo fluindo suavemente.

Diferença dos Loops de Eventos: No Node.js, o loop é otimizado para tarefas de servidor, como acessar arquivos ou bancos de dados, enquanto nos navegadores, ele foca mais em interações com o usuário e renderização de páginas.

**Comment by Mariaeed** [comment]
![Uploading IMG_0142.jpeg…]()

**Comment by PedroLimadaRocha** [comment]
● Runtime :
Processo de execução de um programa/aplicação podendo serem correlacionadas com o
ambiente no qual é executado ou/e tempo da execução do software. Exemplos como JVM
para ambiente de execução e formas de utilizar os processos para desenvolver.
● Node.js :
Motor JavaScript V8 do Chrome, Libuv - Operações assíncronas e Camada C++ - Funções
são os três componentes do node.js.
_ Motor V8:
Ele interpreta seu código em Javascript e traduz para instâncias de objeto em C++ e
executa, assim auxiliando na implementação em todas as funções onde o JavaScript não
tem a capacidade de executar.
_ Libuv:
Sendo uma biblioteca em C++ desenvolvida para node.js inicialmente, porém atualmente
outros tecnologias a utiliza. Focada em assíncronia, onde utiliza thread pools para E/S
permitindo que outras tarefas executem enquanto é esperado o resultado de outras
operações.
_ Assíncronia:
Considera uma técnica onde evita atrasos de processos e esperas desnecessárias, no
qual a realização de um procedimento síncrono há a necessidade de esperar que outros
códigos executem para a continuação do código, sendo um dos melhores processos para
utilização em ambiente Web, assim caso seja necessário a espera se um recurso anterior
prejudicaria a navegação como um todo.
_ Callback:
Um callback é uma função que é passada como parâmetro para outra função e é
executada quando um determinado evento ocorre, ou seja, garante que uma função não
seja executada antes de outra tarefa ser concluída.
_ Loop de Eventos no Node.Js e em Navegadores:
É considerado um loop infinito que espera por tarefas, executa-as e depois espera por
novas .O loop de eventos é único, sem bloqueio e assíncrono. Sendo essenciais nos
processos com operações assíncronas e não bloqueando outros processos.
Na maioria dos Navegadores WEB existe um loop de eventos para cada aba aberta, para
tratar cada processo isoladamente e evitar que uma página com loops infinitos ou com um
excesso de processamento, bloqueie o navegador por completo.

**Comment by devhetor** [comment]
• O que é um runtime
É o tempo de execução do programa, ele pode ser dividido em ambiente de execução(bibliotecas, servicos e ate mesmo o SO) e tempo de execução (período da execução até o encerramento do programa) 

• Componentes do runtime JS Node.js
Motor V8, libuv, camada c++ 

• O que é e o que faz a V8
Ela traduz o código em Java script para c++ para depois enviar esse código para ser compilado 

• O que é e o que faz a Libuv
É uma biblioteca em C que permite o trabalho de entrada e saída de dados assíncronos, o que não bloqueia o Loop principal do node
• O que é assincronia
É uma técnica que permite executar uma função e retornar ela em um outro momento, enquanto executa outras funções. 

• O que são callbacks
É uma função passada por parâmetro para outra função 

• Como funciona o Loop de eventos do Node.js e quais são seus componentes
Ele monitora a ordem da execução das funções, caso encontre uma função assíncrona, ele envia para uma fila (na libuv), a primeira promessa a entrar é a primeira a sair, depois ele retorna o resultado por uma call stack. Os componentes são libuv, call stack, event queue e thread pool 

• Qual a diferença do Loop de Eventos do Node.js para o Loop de eventos utilizados nos Browsers 
No loop de eventos do browser é utilizado para tratar elementos do DOM, como interações do usuário. Enquanto o loop de eventos do node.js é utilizado para operações como acesso ao banco de dados e chamadas externas.

**Comment by JoaoPolloni** [comment]
- Runtime: Ambiente de execução de um programa.
- Componentes do runtime JS Node.js: V8 e Libuv.
- V8: Engine de JavaScript open-source do Google.
- Libuv: Biblioteca multiplataforma para I/O assíncrono.
- Assincronia: Execução não sequencial de operações.
- Callbacks: Funções passadas como argumentos para outras funções, a serem chamadas posteriormente.
Loop de eventos do Node.js: O Loop de Eventos é uma parte essencial do modelo de execução assíncrona do Node.js. Ele gerencia a ordem de execução de operações assíncronas e garante que o código assíncrono seja executado de forma eficiente. O Loop de Eventos do Node.js consiste em três componentes principais:
Call Stack: Mantém o rastreamento das funções que estão sendo executadas no momento.
Callback Queue: Armazena os callbacks das operações assíncronas que foram concluídas e estão prontas para serem executadas.
Event Loop: Responsável por verificar continuamente se há callbacks na Callback Queue e movê-los para a Call Stack para execução, garantindo que o código assíncrono seja executado de forma oportuna.

- Diferença do Loop de Eventos do Node.js para Browsers: Embora o conceito de Loop de Eventos seja semelhante tanto no Node.js quanto nos navegadores web, suas implementações são diferentes devido aos ambientes de execução distintos. No Node.js, o Loop de Eventos é otimizado para operações de I/O assíncronas, como acesso ao sistema de arquivos ou comunicação de rede. Nos navegadores, o Loop de Eventos é otimizado para lidar com eventos de interface do usuário, como cliques do mouse e teclas pressionadas.

**Comment by EuHiroshi** [comment]
1- O que é runtime
	É o período em que um programa está ativo, desde o inicio até quando está encerrado

2- Componentes do runtime JS Node.js
	É uma parte essencial do ambiente de execução que permite que o Node execute código js fora do navegador

3- O que é e o que faz a V8?
	É o motor de js de código aberto desenvolvido pelo Google, e responsável por interpretar e executar o código js.

4- O que é e o que faz a Libuv?
	É um biblioteca em C que permite ao node criar várias operações de entrada e saída de forma assíncrona

5- O que é assincronia?
	Refere-se a operações que não bloqueiam a execução de um programa

6- O que são callbacks?
	São funções que são passadas como argumentos para outras funções e são chamadas de volta após a conclusão

7- Como funciona o Loop de eventos do Node.js e quais são seus componentes?
	Permite a execução de operações de forma não bloqueante lidando com eventos chamados callbacks

8- Qual a diferença do Loop de Eventos do Node.js para o Loop de eventos utilizados nos Browsers?
	O loop de eventos do node é otimizado para operações de entrada e saida intensivas em ambientes de servidor, enquanto o loop de eventos dos navegadores lida com interações do usuário e renderização de páginas web.

**Comment by renan-cardeal2002** [comment]
O que é um runtime?
É um ambiente que permite a execução de programas ou softwares, fornecendo uma camada de abstração entre o aplicativo e o ambiente que será executado. Ele oferece bibliotecas, recursos e componentes importantes pra execução do aplicativo.
Componentes do runtime JS Node.js:
V8 Javascript, Event Loop, Libuv
O que é e o que faz a V8?
O V8 é um motor de execução (tipo um compilador), responsável por interpretar e executar o código Javascript 
O que é e o que faz a Libuv?
A Libuv é uma biblioteca do node que fornece suporte para assincronia, ela implementa um loop de eventos que gerencia assincronia de tarefas, auxiliando em tarefas onde o desenvolvedor precisa agendar eventos, ou uma função deve esperar outra acabar. Usando o setTimeout() podemos agendar uma tarefa e definir um tempo para ela ser executada.
O que é assincronia?
Assincronia é um modo em que os eventos e processos não ocorrem simultaneamente, ou seja, não são sincronizados. Dois eventos podem começar ao mesmo tempo (ou não) e acabar em tempos totalmente diferentes. Isso deixa a manipulação de eventos mais rápida e menos dependentes uma das outras
O que são call-backs?
São funções ou blocos de código, que são passadas como parâmetro, e que podem retornar um valor de uma promessa ou não, são chamadas em determinado momento na função principal.
Como funciona o Loop de eventos do Node.js e quais são seus componentes?
Os eventos e call-backs são registradas na Event Queue, a medida que são executados, eles são colocados da fila de eventos para serem processados, a Thread Pool. O Event Loop monitora constantemente a fila pra verificar se existem eventos para serem processados, e os inicia. C++ Bindings (Vínculos em C++) pega o código do javascript e salva em uma string e fornece bibliotecas com funcionalidades assíncronas pro node. Esse processo continua até que a fila de eventos esteja vazia, momento em que o Node entra em estado ocioso.
Qual a diferença do Loop de Eventos do Node.js para o Loop de eventos utilizados nos Browsers?
O Loop de eventos do Node é baseado em eventos assíncronos, como solicitações de rede, leitura/gravação, e interações com bancos de dados. Já o Loop de eventos dos browsers é focado em eventos do usuário, como cliques, textos escritos, movimentos do mouse e interações, mas também utiliza recursos assíncronos para carregamento de imagens e scripts. Isso significa que o Loop de eventos do node é mais eficiente em operações de entrada e saída, enquanto o dos browsers prioriza uma boa experiência do usuário.

**Comment by SirPripyat** [comment]
Resumo do que se aprendeu em sala - Tópicos de Programação
Leonardo Rossi - 21009699-2

- O que é um runtime? 
 Runtime é o tempo de execução de um software. No contexto de desenvolvimento de software, pode ser relacionado em dois contextos:
  1 - Ambiente de execução: todo conjunto de recursos que são necessários para executar o software 
  2 - Tempo de execução: Tempo em que o software está sendo executado
  
- Quais são os componentes do Node.js?  
Motor V8 do Chrome, Libuv e a camada C++

- O que é o Motor V8? 
 É a engine principal que atua no core do navegador Google Chrome. Ela traduz o código JavaScript para instâncias de objetos C++.
 
- O que faz a Libuv?
  A libuv é uma lib assíncrona desenvolvida na linguagem C que nasceu para dar suporte ao Node.js, porém também dá suporte a outras linguagens. Ela é responsável por lidar com funções como setTimeout.
  
- O que é assíncronia? 
Assíncrona é o ato de executar trechos de código sem travar a execuçãoo de outros trechos de código. Fundamental para fazer requisições http.

- O que são callbacks? 
Callback é uma função que é passada como paramêtro para outra função. Essas funções só são executadas após a conclusão de determinadas tarefas

- Quais são os loops de eventos do Node.js e quais são os seus componentes?
	Os event loop é o que permite com que o Node.Js lide com operações de E/S de forma assíncrona. Os seus principais componentes são event loop, callstack, callstack queue.

**Comment by LuigiCapoia** [comment]
O V8 desempenha um papel importante como mecanismo JavaScript que pega strings JavaScript e como tradução para execução em C++. O libuv  é a biblioteca  em C facilita o suporte multiplataforma para E/S JavaScript assíncrona, o que é fundamental para o funcionamento eficiente do Node.js. O termo runtime do JavaScript refere-se ao ambiente de execução do JavaScript, que inclui mecanismos, pilhas de chamadas, memória e outros componentes que permitem que o código seja executado de forma eficaz em vários contextos.A programação assíncrona ajuda a evitar atrasos ou esperas na execução de programas.Uma função que é passada como parâmetro para outra função e é chamada quando ocorre um determinado evento é conhecido como retorno de chamada.O loop de eventos Node.js usa componentes como retorno de chamada e pilha de chamadas. Isso permite o gerenciamento eficiente de operações assíncronas, tornando o Node.js adequado para aplicativos escalonáveis ​​orientados a eventos.O loop de eventos no Node.js e nos navegadores lidam com eventos assíncronos, mas diferem na aplicação uma vez que o Node.js usa libuv para lidar com código assíncrono.

**Comment by CauaMurata** [comment]
O que é runtime?
É o tempo em que o código é executado até o momento em que é encerrado. Durante esse período, o programa realiza tarefas interagindo com o ambiente, manipula dados, acessa recursos e etc.

Runtime JS node.js possui 3 componentes:
1-Motor JavaScript v8 do Chrome.
2-libuv - operações assíncronas.
3-camada C++ funções para ajudar o fluxo de dados.

O motor v8 interpreta e executa o código JavaScript, traduzindo suas instâncias de objeto em C++.

Libuv é uma biblioteca C que permite o node criar: Timers, Threads, Child process, Async functions e etc.

A programação assíncrona permite que seu programa inicie uma tarefa de longa duração e ainda seja capaz de responder a outros eventos enquanto essa tarefa é executada, em vez de ter que esperar até que essa tarefa seja concluída.

Callback permite passar uma função como argumento para outra função, que será executada em um momento específico ou em resposta a um evento, ou seja, é uma forma de delegar a execução de um trecho de código para outra função.

O loop de eventos permite lidar com eventos assíncronas de E/S, aguardando e processando eventos em um ciclo contínuo. Seus componentes são: event queue e Threads pool.

Uma das diferenças é que o node utiliza o modelo single-thread enquanto que navegadores utilizam multi-threaded.

**Comment by renandias26** [comment]
1 - O que é um runtime
    É o ambiente onde o JS é executado, fornecendo funcionalidades extras de acordo com o ambiente.
        - Node é um ambiente que fornece funcionalidades server-side
            - Manipular arquivos, sistema, request HTTP
        - Browser é um ambiente que fornece funcionalidades client-side
            - Funcionalidades como Web API
                - DOM
                - Fetch API
2 - Componentes do runtime JS Node.js
    Core API, C++ bindings, V8, Libuv
3 - O que é e o que faz a V8
    Ele executa o código JS, e o traduz para C. Permite criar funções personalizadas, fornecendo funções que não estão nativamente no node.js
4 - O que é e o que faz a Libuv
     Libuv é uma biblioteca em C++ que permite a manipulação de eventos assíncronos
5 - O que é assincronia
    Assincronia é a possibilidade de a execução do seu código não ficar travada em uma tarefa que demora para ser executada.
    Ela permite que durante a execução de uma tarefa que demora para ser finalizada, o restante do código continua executando enquanto essa tarefa está sendo finalizada.
6 - O que são callbacks
    São funções executadas por outras funções
7 - Como funciona o Loop de eventos do Node.js e quais são seus componentes
    Ele funciona como um loop infinito que, a cada iteração, verifica se há eventos a serem executados na Task Queue, um de seus componentes além de outros como, call stack, Event Loop, Timers, Node APIs, entre outros
8 - Qual a diferença do Loop de Eventos do Node.js para o Loop de eventos utilizados nos Browsers
    O loop de eventos no node possui fases específicas para lidar com cada tipo de tarefa, já no browser a divisão das tarefas é feito apenas entre macro e micro
    Enquanto o node fornece API para operações não bloqueantes, o browser fornece operações ligadas a interações do usuário como um clique de botão

**Comment by karenreisss** [comment]
O que é um runtime?
é dividido em duas principais interpretações:
1. ambiente de execução: é o conjunto de recursos necessários para que o programa seja executado (máquina virtual, biblioteca, sistema operacional…).
2. ⁠tempo de execução: período em que o programa está em execução, desde o momento que inicia até o momento que encerra.
durante o tempo de execução o programa interage com o ambiente de execução para realizar o que está sendo solicitado nele

Componentes do runtime JS Node.js?
principais componentes do runtime JS Node.js são:
• motor JavaScript V8 do Chrome
• Libuv
• camada C++

O que é e o que faz a V8?
V8 é a engine do Google que permite que código javascript seja interpretado e executado. Pega uma string com código javascript, valida, e o executa em C++.

O que é e o que faz a Libuv?
 é uma biblioteca em C++ desenvolvida para Node.js e atualmente outras tecnologias a utilizam. Focada em assíncronia, utiliza thread pools para entrada e saída permitindo que outras tarefas executem enquanto é esperado o resultado de outras operações.

O que é assincronia?
são eventos e processos nao ocorrem simultaneamente, não são sincronizados.

O que são callbacks?
são funções passadas como parâmetros para outras funções, para que sejam executadas após a conclusão de alguma operação, sendo um fluxo de trabalho assincrono. 

Como funciona o Loop de eventos do Node.js e quais são seus componente?
quando o loop de eventos entra em uma determinada fase, ele opera callbacks nessa fila de fase até que a fila se esgote e o numero máximo de callbacks seja executado, então passa para a próxima fase. seus principais componentes são: call stack, callback queue e event loop.

**Comment by veigasdevCesu** [comment]
O que é um runtime?

é o período que um programa está em execução, desde o momento que é iniciado até o momento que é encerrado;

Componentes do runtime JS NODE.js

V8, Libuv, módulos essenciais e de terceiros

O que é e o que faz a V8

É um motor de execução do JS desenvolvido pelo Google. Ele converte o código JS em código de maquina.

O que é e o que faz a Libuv

É uma biblioteca utilizada pelo NODE.js para operações assíncronas e manipulação de eventos de sistema. Permite criar threads, timers, child processes, event loops e async functions

O que é assincronia

É um elemento que permite que as tarefas sejam executavas de forma independente, sem esperar pelo final de alguma operação para continuar

O que são callbacks

São funções passadas como argumentos para outras funções, executadas depois de uma função assíncrona ser concluida.

Como funciona o loop de eventos do NODE.js e quais são seus componentes

O loop de ebentos do Node.js é essencial para lidar com operações assíncronas de forma eficiente. Ele opera de maneira cíclica, verificando continuamente a presença de eventos na fila de eventos e executando os callbacks associados. As operações assíncronas, como solicitações de rede ou leitura de arquivos, são tratadas por meio de callbacks na fila de callbacks. O Node.js utiliza um pool de threads para operações de E/S intensivas, garantindo que o loop principal não seja bloqueado. Esse sistema permite que o código JavaScript seja executado de forma não bloqueante, garantindo alta eficiência no tratamento de operações assíncronas.

Qual a diferença do loop de eventos do node.js para o loop de eventos utilizados nos browsers

- O loop de eventos no Node.js e nos navegadores é usado para lidar com operações assíncronas, mas existem diferenças significativas em suas implementações e funcionalidades. No Node.js, o foco está em operações de I/O de sistrma, como acesso a arquivos e rede, adequado para aplicativos do lado do servidor. Nos navegadores, o foco é em interações com o DOM e eventos de usuário, para aplicativos do lado do cliente. Cada ambiente oferece APIs específicas para suas respectivas tarefas, adaptadas às necessidades do desenvolvimento de aplicativos em seus contextos específicos.

**Comment by felipedefendi** [comment]
1- O que é um runtime:

Um runtime é como um ambiente onde o código de um programa é executado. Ele inclui tudo o que é necessário para que o código funcione corretamente.

2- Componentes do runtime JS Node.js:

O Node.js, como um ambiente de execução para JavaScript, tem dois componentes principais: o motor V8, que executa o código JavaScript, e a biblioteca Libuv, que ajuda na execução de operações de entrada e saída de forma eficiente.

3- O que é e o que faz a V8:

A V8 é como o "cérebro" por trás do Node.js. Ele é responsável por entender e executar o código JavaScript que você escreve.

4- O que é e o que faz a Libuv:

A Libuv ajuda o Node.js a lidar com tarefas como ler arquivos, acessar a internet e outras operações que podem levar algum tempo. Ela garante que essas tarefas sejam realizadas de forma rápida e eficiente.

5- O que é assincronia:

Assincronia é quando o código continua executando enquanto espera por algo, como ler um arquivo da internet. Isso é útil para que o programa não fique parado, esperando que uma tarefa seja concluída.

6- O que são callbacks:

Callbacks são como notas que você deixa para o Node.js fazer algo quando uma determinada tarefa é concluída. Por exemplo, você pode pedir para o Node.js ler um arquivo e, quando terminar, ele chama a função que você deixou como "callback".

7- Como funciona o Loop de eventos do Node.js e quais são seus componentes:

O Loop de Eventos do Node.js é como um gerente que organiza as tarefas do programa. Ele verifica constantemente se há algo a ser feito e processa as tarefas conforme necessário.

8- Qual a diferença do Loop de Eventos do Node.js para o Loop de eventos utilizados nos Browsers:

O Loop de Eventos do Node.js e o dos navegadores têm trabalhos semelhantes, mas são adaptados para ambientes diferentes. Enquanto o Loop do Node.js lida principalmente com tarefas de servidor, como ler arquivos e acessar bancos de dados, o Loop dos navegadores se concentra em interações do usuário, como clicar em botões e preencher formulários.

**Comment by jvbt3** [comment]
[atividade.pdf](https://github.com/TI-UNICESUMAR/2024-topicos-em-programacao-ESOFT7S/files/14305261/atividade.pdf)
